### DISCUSSION TITLE
the middleware is so slow

### DISCUSSION THREAD

**Discussion by wesdu** [fastapi/fastapi#6985] [Questions] [answered]
I using the test case which is a helloworld example with 6 empty middleware(doing nothing).
And it significant slowdown. Loss 70% performance vice not using middleware.
And i notice the code using in TechEmpower is without middleware.
Deep in the code I found out the middleware runing is using too many coroutine(create task).
I think there should be some performance fixed for the middleware in fastapi.

**Comment by Mause** [comment]
Can you give a demonstration of your issue?

**Comment by wesdu** [comment]
```
from fastapi import FastAPI, Request
import uvicorn

app = FastAPI(debug=False)


@app.middleware("http")
async def a(request: Request, call_next):
    response = await call_next(request)
    return response

@app.middleware("http")
async def b(request: Request, call_next):
    response = await call_next(request)
    return response

@app.middleware("http")
async def c(request: Request, call_next):
    response = await call_next(request)
    return response

@app.middleware("http")
async def d(request: Request, call_next):
    response = await call_next(request)
    return response


@app.get("/")
async def run():
    return {"message": "hello"}


if __name__ == '__main__':
    uvicorn.run('test:app', host='0.0.0.0', port=8888, reload=True, debug=False)
```
when using the 4 level middleware 
I got 
`Requests per second:    1753.81 [#/sec] (mean)`
 using `ab -c 100 -n 10000`

otherwise, remove the empty middleware 
```
from fastapi import FastAPI, Request
import uvicorn

app = FastAPI(debug=False)

@app.get("/")
async def run():
    return {"message": "hello"}


if __name__ == '__main__':
    uvicorn.run('test:app', host='0.0.0.0', port=8888, reload=True, debug=False)
```
I got 
`Requests per second:    3450.91 [#/sec] (mean)`

**Comment by wesdu** [comment]
The middleware has a very strong impact on the performance, I don't think it should be.

**Comment by ycd** [comment]
I made a little profiling on this, here are my findings:

I ran 1000 request with the code you given above.

```
Summary:
  Total:	5.5514 secs
  Slowest:	0.3520 secs
  Fastest:	0.1784 secs
  Average:	0.2755 secs
  Requests/sec:	180.1345

  Total data:	19000 bytes
  Size/request:	19 bytes

Response time histogram:
  0.178 [1]	|
  0.196 [12]	|■■
  0.213 [17]	|■■
  0.231 [15]	|■■
  0.248 [154]	|■■■■■■■■■■■■■■■■■■■■■
  0.265 [292]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.283 [240]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.300 [26]	|■■■■
  0.317 [54]	|■■■■■■■
  0.335 [103]	|■■■■■■■■■■■■■■
  0.352 [86]	|■■■■■■■■■■■■
```

So we averaged **0.275 ms** per request.

For 1000 request, we used the the middleware  `Request Count * 4` = **4000** times.

For 1000 request which took 5.5 seconds to complete. Middleware used 1.6 second of that time. Which is nearly 29%.

```
name                                  ncall  tsub      ttot      tavg      
...py:19 BaseHTTPMiddleware.__call__  4000  0.042633  1.170604  0.000292
..py:28 BaseHTTPMiddleware.call_next  4000   0.062403  0.393508  0.000098
```

So i'm not sure, is there a bottleneck or not because middlewares are generally expensive operations.

**Comment by wesdu** [comment]
> I made a little profiling on this, here are my findings:
> 
> I ran 1000 request with the code you given above.
> 
> ```
> Summary:
>   Total:	5.5514 secs
>   Slowest:	0.3520 secs
>   Fastest:	0.1784 secs
>   Average:	0.2755 secs
>   Requests/sec:	180.1345
> 
>   Total data:	19000 bytes
>   Size/request:	19 bytes
> 
> Response time histogram:
>   0.178 [1]	|
>   0.196 [12]	|■■
>   0.213 [17]	|■■
>   0.231 [15]	|■■
>   0.248 [154]	|■■■■■■■■■■■■■■■■■■■■■
>   0.265 [292]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
>   0.283 [240]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
>   0.300 [26]	|■■■■
>   0.317 [54]	|■■■■■■■
>   0.335 [103]	|■■■■■■■■■■■■■■
>   0.352 [86]	|■■■■■■■■■■■■
> ```
> 
> So we averaged **0.275 ms** per request.
> 
> For 1000 request, we used the the middleware `Request Count * 4` = **4000** times.
> 
> For 1000 request which took 5.5 seconds to complete. Middleware used 1.6 second of that time. Which is nearly 29%.
> 
> ```
> name                                  ncall  tsub      ttot      tavg      
> ...py:19 BaseHTTPMiddleware.__call__  4000  0.042633  1.170604  0.000292
> ..py:28 BaseHTTPMiddleware.call_next  4000   0.062403  0.393508  0.000098
> ```
> 
> So i'm not sure, is there a bottleneck or not because middlewares are generally expensive operations.

You should add the "Concurrency" scene in your testing. Because the the problem is too many coroutine created which impact the  Concurrency ability.
The middleware code detail in the framework is about "asyncio queue" and "create task".

**Comment by ycd** [comment]
@wesdu i ran with 50 concurrency.

**Comment by Kludex** [comment, marked as answer]
`BaseHTTPMiddleware` should be avoided (but not related to this issue). 

Can you `app.add_middleware(UsefulMiddleware)` and check with this one?

```python
class UsefulMiddleware:
    def __init__(self, app):
        self.app = app
        
    async def __call__(self, scope, receive, send):
        await self.app(scope, receive, send)
```

**Comment by wesdu** [comment]
@Kludex Yep, the callback way is fast.
But not so “Useful” -- Fast to run but not so easy to write.
There should be a better `BaseHTTPMiddleware `.

**Comment by ycd** [comment]
> But not so “Useful” -- Fast to run but not so easy to write.
> There should be a better BaseHTTPMiddleware 

What do you mean? There should be an easier way or faster way?

**Comment by wesdu** [comment]
> > But not so “Useful” -- Fast to run but not so easy to write.
> > There should be a better BaseHTTPMiddleware
> 
> What do you mean? There should be an easier way or faster way?

both

**Comment by ycd** [comment]
We are always open to new ideas, feel free to open a PR if you have a better implementation idea!

**Comment by wesdu** [comment]
conclusion: Don't use BaseHTTPMiddleware. It loses performance and even context.

**Comment by tomeras1** [comment]
@wesdu did u eventually used UsefulMiddleware variation? I also ran into performance issues

**Comment by wesdu** [comment]
@tomeras1 Just write the middleware in the ASGI classic way. You can take the build-in middleware in the source code as example.

**Comment by nisarg-soni-ss** [comment]
There is different scenario in my case. When my app is in load, it turns out the requests are taking soo much time sometimes more than 10 sec. the same req takes 500 ms when not in load.

So we Debug that issue and put enter exit logs with request ids at router and middleware level, which turns out that the response has been returned from the router but middleware exit log was after 8-9 secs. which was taking so much time. and in between that 8-9 secs the app was handling other request. so even after completion of execution app was not able to return the response.

i guess middleware puts all the request to eventloop and there can be some issue in that, I can be wrong. but this was heppening

**Comment by ClaytonJY** [comment]
I don't have precise performance numbers, but I can confirm that the "ASGI classic way" seems to be more performant than the decorator approach. It was a lot trickier to write, but Starlette and ASGI are old enough that ChatGPT (4) was quite helpful.

Given some of the reports above, and a recommendation from a maintainer to stay away from the decorator, should there be documentation added to warn users about this trap in performance-sensitive scenarios?

**Comment by aldem** [comment]
While the issue is quite old, it is still present in the most recent version - just encountered it (FastAPI 0.110.1).

The problem is that any kind of BaseHTTPMiddleware is terribly slow - be it a decorator or a class, in my test case just single, no-op middle-ware reduces RPS _five(!) times_, if I add the second one - it gives _additional_ 2x slowdown.

In my test (Python 3.11), with empty request handler, I get:
- ~ 50K RPS (2 cores, 2 workers, 512 concurrent requests)
- ~ 10K RPS after adding one no-op middle-ware 
- ~ 5K RPS and after adding the second one

With single worker it is "just" _3.6 times_ slower - ~ 22K RPS without any middle-ware and ~ 6K RPS with no-op one.

On a single-core this basically means that once I have around 6K RPS I have no CPU time left for anything else. But, I tested this on a high-end desktop CPU (Ryzen 5950), which means that on a typical server, inside a container or VM it will be around 3K RPS/core - and only overhead.

ASGI-level middle-ware is fast enough but it is a bad option - I have to parse headers (so the work is done twice), don't get the signature of the handler and could not use DI or anything else that is typically bound to the request, not without nasty workarounds at least.

In fact, this is a big show stopper - while FastAPI has a lot of nice features (I really like it), this one problem is ruining everything - I could not use it without middle-ware but performance impact is unbearable, even if I combine all processing into single middle-ware.

And since I don't have this issue with Sanic or BlackSheep (where request-level middle-ware overhead is nearly invisible - around 2-3%), I believe that it could be made way faster.

Any ideas why this problem was not addressed at all? Or the problem is actually in Starlette and not in FastAPI?

**Comment by ClaytonJY** [comment]
For the observers, I'd like to highlight the much-improved Starlette docs, quite helpful for writing pure ASGI middleware when appropriate: https://www.starlette.io/middleware/#pure-asgi-middleware

There's also a brief mention of limitations to `BaseHTTPMiddleware`, though it does not mention the performance aspect.
